### PR TITLE
multi layer cache

### LINKED ISSUE #10: Multi layered caching
Searching for cached markov models should work something like this:

1. do I have this key in memory?  (ignore in dev)
2. is fs caching enabled (might be disabled in dev)? do I have this key in fs? no?
3. do I have valid blob store creds? do I have this key in {{ remote key/value store }}? no?
4. okay regenerate model

Create following "cache backends":

- memory
- filesystem
- blob storage

Check in each cache backend when looking for a cached value

### PR DESCRIPTION
**Pull request by ridhoq**
Implements a caching system that allows for numerous backends. If there's a cache miss on a backend and a cache hit on a subsequent backend, we'll cache the value in the prior caches for faster retrieval next time.

When we run `dcr cli d thomas austin thomas d` we should see the first retrievals pulling from the blob cache. Subsequent runs will be pulling from the LFU (in memory) cache.

Closes #10.

### CODE REVIEW

File: ditto-markov/README.md [hunk 1]
```diff
@@ -4,4 +4,4 @@ Rewrite of the markov chain functionality into a python service using a Cosmos d
 ## Getting Started
 Currently, there's just a CLI wrapper to generate the user's chat messages
 1. copy `.env.example` to `.env` and drop in the correct values
-1. `docker-compose run --rm cli <slack-user>` will return 20 messages that sound like the user
\ No newline at end of file
+1. `docker-compose run --rm cli <slack-user>` will return 20 messages that sound like the user. Optionally, pass a space delimited list of users
```
Comment (austinpray) [inline ditto-markov/README.md:7]: :+1:

File: ditto-markov/ditto_markov/cache/cache.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+import typing
+from typing import Union
+
+import cachetools
+
+
+class Cache(ABC):
```
Comment (austinpray) [inline ditto-markov/ditto_markov/cache/cache.py:9]: nice

File: ditto-markov/ditto_markov/cache/cache.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+RegisterableCache = Union[Cache, cachetools.Cache]
+
+
+class CacheRepository(Cache):
+    caches: typing.OrderedDict[str, RegisterableCache]
+
+    def __init__(self):
+        self.caches = OrderedDict()
```
Comment (austinpray) [inline ditto-markov/ditto_markov/cache/cache.py:30]: In recent versions of py dicts are ordered by default I'm pretty sure

File: ditto-markov/docker-compose.yaml [hunk 4]
```diff
@@ -30,8 +36,9 @@ services:
       COSMOS_ACCOUNT_URI: https://cosmosdb:3000
       COSMOS_DATABASE_ID: ditto-test
       COSMOS_DISABLE_TLS: "True"
+      # default conn str from https://github.com/Azure/Azurite#http-connection-strings
```
Comment (austinpray) [inline ditto-markov/docker-compose.yaml:39]: scared me at first but :+1:

**Review by austinpray — APPROVED** [review]
image

get some warnings in the tests

Otherwise werkin great for multiple runs

image

**Comment by ridhoq** [comment]
> get some warnings in the tests

I think this has to do with not mounting `pytest.ini` - any ideas on why that might be?

**Comment by ridhoq** [comment]
> get some warnings in the tests

@austinpray I think this has to do with not mounting `pytest.ini` - any ideas on why that might be?

File: ditto-markov/ditto_markov/cache/cache.py (hunk 3 above)
Comment (ridhoq) [reply ditto-markov/ditto_markov/cache/cache.py:30]: I started with an `OrderedDict`, realized that dicts are now ordered, and then decided to go back to `OrderedDict` just to make everything explicit.

Imagine having decades and decades of language behavior assumptions shattered

**Pull request merged**